eventsources/sources/calendar/start.go

```diff
@@ -88,45 +173,82 @@ func (el *EventListener) StartListening(ctx context.Context, dispatch func([]byt
 		return nextT
 	}
 
-	lastT := time.Now()
+	lastT, err := el.getExecutionTime()
+	if err != nil {
+		return err
+	}
+
 	var location *time.Location
 	if calendarEventSource.Timezone != "" {
-		log.Info("loading location for the schedule...", zap.Any("location", calendarEventSource.Timezone))
+		el.log.Info("loading location for the schedule...", zap.Any("location", calendarEventSource.Timezone))
 		location, err = time.LoadLocation(calendarEventSource.Timezone)
 		if err != nil {
 			return errors.Wrapf(err, "failed to load location for event source %s / %s", el.GetEventSourceName(), el.GetEventName())
 		}
 		lastT = lastT.In(location)
 	}
+	sendEventFunc := func(tx time.Time) error {
+		eventData := &events.CalendarEventData{
+			EventTime:   tx.String(),
+			UserPayload: calendarEventSource.UserPayload,
+			Metadata:    calendarEventSource.Metadata,
+		}
+		payload, err := json.Marshal(eventData)
+		if err != nil {
+			el.log.Error("failed to marshal the event data", zap.Error(err))
+			// no need to continue as further event payloads will suffer same fate as this one.
+			return errors.Wrapf(err, "failed to marshal the event data for event source %s / %s", el.GetEventSourceName(), el.GetEventName())
+		}
+		el.log.Info("dispatching calendar event...")
+		err = dispatch(payload)
+		if err != nil {
+			el.log.Error("failed to dispatch calendar event", zap.Error(err))
+			return errors.Wrapf(err, "failed to dispatch calendar event")
+		}
+		if el.eventPersistence != nil && el.eventPersistence.IsEnabled() {
+			event := persist.Event{EventKey: el.getPersistenceKey(), EventPayload: string(payload)}
+			err = el.eventPersistence.Save(&event)
+			if err != nil {
+				el.log.Error("failed to persist calendar event", zap.Error(err))
+			}
+		}
+		return nil
+	}
 
+	el.log.Info("Calendar event start time:", zap.Any("Time", lastT.Format(time.RFC822)))
 	for {
 		t := next(lastT)
+
+		// Catchup scenario
+		// Trigger the event immediately if the current schedule time is earlier then
+		if time.Now().After(t) {
+			el.log.Info("triggering catchup events", zap.Any(logging.LabelTime, t.UTC().String()))
+			lastT = t
```

Comment:
`lastT = t` should be after successfully `sendEventFunc(t)`

Reply:
Good catch